README
 [x] FUNCTIONS PROGRESS TESTING

Made by Edoardo Mantovani, 2020

function list with working grades:

W == work without any error

C == Crash

F == Fail unconditionally

N == Not Tested

=================================
FUNCTION LIST:
=================================

lorcon_get_error   (W)
lorcon_get_error   (W)
lorcon_find_driver (W)
lorcon_auto_driver (W)

lorcon_free_driver_list (N)

lorcon_create       (W)
lorcon_free         (W)
lorcon_set_timeout  (W)
lorcon_get_timeout  (W)
lorcon_open_inject  (W) # note: this works under some devices (tuntap)  but it crash under madwifing driver
lorcon_open_monitor (W) # note: same as upside 
lorcon_open_injmon  (W) # note: same as upside

# for now everything works until line 102 of lorcon2.xs

lorcon_set_vap         (W) 
lorcon_get_vap         (W)
lorcon_get_capiface    (W)
lorcon_get_driver_name (W)
lorcon_close           (W)
lorcon_get_datalink    (F)
lorcon_set_datalink    (F)
lorcon_set_channel     (F)
lorcon_get_channel     (W)

# some errors occoured, line 141 of lorcon.xs

lorcon_get_hwmac         (N)
lorcon_set_hwmac         (N)
lorcon_get_pcap          (?) # note: it doesn't fail but it gives no output
lorcon_get_selectable_fd (F)
lorcon_next_ex           (N)
lorcon_set_filter        (?) # note: don't know what filters are used
lorcon_breakloop         (W)
lorcon_inject            (W)
lorcon_send_bytes        (W)
lorcon_get_version       (W)

# some fails occoured, line 202 of lorcon.xs tested

lorcon_add_wepkey            (N)
lorcon_set_useraux           (N)
lorcon_get_useraux           (C)
lorcon_packet_free           (C)
lorcon_packet_decode         (W)
lorcon_packet_set_channel    (~) # note: sometimes it crash and sometimes no
lorcon_packet_from_dot3      (N)
lorcon_packet_to_dot3        (N)
lorcon_ifup                  (F)
lorcon_packet_get_source_mac (N)
lcpf_randmac                 (..)
lorcon_packet_get_dest_mac   (..)
lorcon_packet_get_bssid_mac  (..)
lorcon_ifdown                (F)
lorcon_set_complex_channel   (N)
lorcon_get_complex_channel   (N)
lorcon_parse_ht_channel      (N)
lorcon_multi_create          (W)
lorcon_multi_free            (C)
lorcon_multi_add_interface   (W)
lorcon_multi_del_interface   (N)
lorcon_multi_get_interfaces  (W)
lorcon_multi_get_next_interface
lorcon_multi_interface_get_lorcon (C)
drv_madwifing_init           (W)
lorcon_airjack_init          (?)
# aj...
#nl80211..
nl80211_find_parent          (W)

# iwconfig...

TODO:
# test until line 500

drv_file_listdriver          (W)


perl/Changes
// Made by Edoardo Mantovani, 2020

================================================
CHANGED PROJECT, NOW IT WILL BE AIR::LEGACY
================================================

00.00   Tue  Nov 05 21:00:00 CET 2020
  - fixed small bugs
  - amplied documentation
  - added 2 new functions: send_generic_packet and send_packet_internal
  - first inclusion of WPA/WPA2 protocols inside Lorcon2
  - updated new test: 25-test-reaver.t
  - added new suggested perl library: Linux::TunTap
  - start works for a more advanced tun/tap interface
  - first beta for wireless driver writing
  - added new wireless drivers documentation
  - added new test : 26-test-reaver-xml.t
  - added new packet crafting functions from the Lorcon2 library: 
     * lcpf_assocresp
     * lcpf_data
     * lcpf_qosheaders
  - fixed small bugs
  - added "pixie_run" and "pixie_thread" functions
  - added airpcap files
  - created official site
  - added wps_send.pl example: show how to send WPS packets
  - added packet_analysis.pl example: show how to examine, convert and entropize captured packets
  - added wps_to_json.pl example: show how to convert WPS stats into json format
  - fixed broken examples
  - added new suggested library: Net::Wireless::802_11::WPA::CLI
  - included unistd.h header, usefull for check if file exist 
  - added wi_ related functions
    * wi_read
    * wi_write
    * wi_set_channel
    * wi_get_channel
    * wi_get_freq
    * wi_set_freq
    * wi_get_monitor
    * wi_get_ifname
    * wi_close
   - added OSdep tag
   - added new function: tx80211_decodepacket 
   - added new function: apitest_packet_hdlr
   - added new suggested library: IO::Socket::SSL::Intercept
   - fixed memory leak bugs
   - added aircrack-ng  interface
   - added aircrack-ng installer
   - added ac_crypto_engine functions and related
   - added ETHERNET constant class [+182]
   - deleted useless headers
   - updated official site
   - fixed compilation errors, now functions works fine
   - toggled perl shellbag: extended portability of the entire project
   - deleted useless python's PyLorcon2 scripts
   
Revision history for the (old)  Perl extension Air::Lorcon2.


29.00   Sun Opt 18 16:00:00 CET 2020
  - fixed small bugs
  - added new FAQ
  - added new documentation links
  - updated POD documentation
  - updated every file to version 29.00
  - corrected grammar spellings and other programming errors in the README.md file 
  - added first version of cfg80211 header
  - updated Devel::PPPort to version 3.62
  - added few new prototypes for future perl cfg80211 control
     * convert_key_from_CPU
     * send_key_to_dongle
     * brcmf_cfg80211_request_ap_if
     * brcmf_mon_add_vif
     * brcmf_cfg80211_add_iface
     * brcmf_notify_escan_complete
  - fixed association_reason_codes_HASH and dissociation_reason_codes_HASH functions
  - officially added pure-perl function: FindLinkage
  - added function tx80211_meta: this will initialize a tx80211 structure into perl execution
  - added new function: tx80211_free
  - added new function: tx80211_packet_meta 
  - toggled Data::Dumper requirement and added test for:
    * tx80211_meta
    * tx80211_packet_meta
    * tx80211_free
    * tx80211_setmode
    * tx80211_geterrstr
    * tx80211_getmode
    * tx80211_gettxrate
  - added new test: test-21-wtinj.t
    tests:
    * wtinj_open  
    * wtinj_close
    * wtinj_setchannel
    * wtinj_getchannel
    * wtinj_setmode
    * wtinj_getmode
  - added 2 new functions: 
    * filter_association_codes
    * filter_dissociation_codes
  - fixed tx80211_decode.c file and added new function: tx80211_decodepkt
  - added new test: 22-test-ajinj.t
  - officially added ajinj_close and ajinj_open functions 
  - added "-oS" optimization flag
  - started first stage for Air::Lorcon2::Extra
  - added new function class: crypto
    * sha1_process
    * sha1_update
    * sha1_finish
    * sha1_starts
    * sha1_hmac_starts
    * sha1_hmac_update
    * sha1_hmac_finish
    * sha1_hmac
  - added new function: sha1_meta (spawn a sha1_context object into perl code)
  - added new function: sha1_hmac_meta (spawn a sha1_hmac_context object into perl code)
  - added new test: 20-test-hash.t ( tests some hashing functions )
  - fixed tests 20 and 23
  - started the Air::Reaver inclusion inside code 
    added GLOB struct, usefull for the following functions:
      * build_radio_tap_header
      * get_wps
      * get_ap_capability
      * set_channel
      * get_channel
      * set_bssid
      * build_wps_probe_request
      * build_association_management_frame
      * build_authentication_management_frame
      * build_snap_packet
      * build_eap_packet
   - added new functions for spawning perl's MANAGEMENT structs:
      * auth_management_meta
      * beacon_management_meta
      * assoc_response_meta
      * assoc_request_meta
   - added new constant class: wps_constant 
   - added 2 new functions: 
      * globule_init 
      * globule_deinit
   - added 5 new reaver function for packet crafting:
      * build_dot1X_header
      * build_eapol_start_packet
      * build_eap_failure_packet
      * build_tagged_parameter
      * build_wps_tagged_parameter
   - first version of "reaver_inject" function, the contropart of lorcon_inject
   - added new function tag: packet_checksum, which offers functions for mathematical verification of captured packets
      * packet_crc
      * packet_entropy
   - added new test: 24-test-checksum.t (test packet checksum capabilities of Air::Lorcon2)
   - officially added "wps_data_to_json" function under :reaver export tag


  
28.00   Sun Opt 04 16:00:00 CET 2020
  - toggled donation tag and modified markdown graphic
  - fixed ALL tests, now the "make full" installation works without any problems under Ubuntu os
  - updated 01-test-auto.t: rewritten using Test::More
  - updated 02-test-consts.t: rewritten using Test::More
  - updated 03-test-detection.t: rewritten using Test::More

C/sha1.c
/*
 *  FIPS-180-1 compliant SHA-1 implementation
 *
 *  Copyright (C) 2001-2003  Christophe Devine
 *  Copyright (C) 2012       Chris Elston, Katalix Systems Ltd <celston@katalix.com>
 *
 *  Ported into Lorcon, dragorn@kismetwireless.net
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 *
 *  Changed to use guint instead of uint 2004 by Anders Broman
 *  Original code found at http://www.cr0.net:8040/code/crypto/sha1/
 *  References: http://www.ietf.org/rfc/rfc3174.txt?number=3174
 *
 *  2012-08-21 - C Elston - Split sha1_hmac function to allow incremental usage.
 */

#include "sha1.h"
#include <string.h>

#define GET_UINT32(n,b,i)                       \
{                                               \
    (n) = ( (uint32_t) (b)[(i)    ] << 24 )       \
        | ( (uint32_t) (b)[(i) + 1] << 16 )       \
        | ( (uint32_t) (b)[(i) + 2] <<  8 )       \
        | ( (uint32_t) (b)[(i) + 3]       );      \
}

#define PUT_UINT32(n,b,i)                       \
{                                               \
    (b)[(i)    ] = (uint8_t) ( (n) >> 24 );       \
    (b)[(i) + 1] = (uint8_t) ( (n) >> 16 );       \
    (b)[(i) + 2] = (uint8_t) ( (n) >>  8 );       \
    (b)[(i) + 3] = (uint8_t) ( (n)       );       \
}


void sha1_starts( sha1_context_t *ctx )
{
    ctx->total[0] = 0;
    ctx->total[1] = 0;

    ctx->state[0] = 0x67452301;
    ctx->state[1] = 0xEFCDAB89;
    ctx->state[2] = 0x98BADCFE;
    ctx->state[3] = 0x10325476;
    ctx->state[4] = 0xC3D2E1F0;
}

void sha1_process( sha1_context_t *ctx, const uint8_t data[64] )
{
    uint32_t temp, W[16], A, B, C, D, E;

    GET_UINT32( W[0],  data,  0 );
    GET_UINT32( W[1],  data,  4 );
    GET_UINT32( W[2],  data,  8 );
    GET_UINT32( W[3],  data, 12 );
    GET_UINT32( W[4],  data, 16 );
    GET_UINT32( W[5],  data, 20 );
    GET_UINT32( W[6],  data, 24 );
    GET_UINT32( W[7],  data, 28 );
    GET_UINT32( W[8],  data, 32 );
    GET_UINT32( W[9],  data, 36 );
    GET_UINT32( W[10], data, 40 );
    GET_UINT32( W[11], data, 44 );
    GET_UINT32( W[12], data, 48 );
    GET_UINT32( W[13], data, 52 );
    GET_UINT32( W[14], data, 56 );
    GET_UINT32( W[15], data, 60 );

#define S(x,n) ((x << n) | ((x & 0xFFFFFFFF) >> (32 - n)))

#define R(t)                                            \
(                                                       \
    temp = W[(t -  3) & 0x0F] ^ W[(t - 8) & 0x0F] ^     \
           W[(t - 14) & 0x0F] ^ W[ t      & 0x0F],      \
    ( W[t & 0x0F] = S(temp,1) )                         \
)

#define P(a,b,c,d,e,x)                                  \
{                                                       \
    e += S(a,5) + F(b,c,d) + K + x; b = S(b,30);        \
}

    A = ctx->state[0];
    B = ctx->state[1];
    C = ctx->state[2];
    D = ctx->state[3];
    E = ctx->state[4];

#define F(x,y,z) (z ^ (x & (y ^ z)))
#define K 0x5A827999

    P( A, B, C, D, E, W[0]  );
    P( E, A, B, C, D, W[1]  );
    P( D, E, A, B, C, W[2]  );
    P( C, D, E, A, B, W[3]  );
    P( B, C, D, E, A, W[4]  );
    P( A, B, C, D, E, W[5]  );
    P( E, A, B, C, D, W[6]  );
    P( D, E, A, B, C, W[7]  );
    P( C, D, E, A, B, W[8]  );
    P( B, C, D, E, A, W[9]  );
    P( A, B, C, D, E, W[10] );
    P( E, A, B, C, D, W[11] );
    P( D, E, A, B, C, W[12] );
    P( C, D, E, A, B, W[13] );
    P( B, C, D, E, A, W[14] );
    P( A, B, C, D, E, W[15] );
    P( E, A, B, C, D, R(16) );
    P( D, E, A, B, C, R(17) );
    P( C, D, E, A, B, R(18) );
    P( B, C, D, E, A, R(19) );

#undef K
#undef F

#define F(x,y,z) (x ^ y ^ z)
#define K 0x6ED9EBA1

    P( A, B, C, D, E, R(20) );
    P( E, A, B, C, D, R(21) );
    P( D, E, A, B, C, R(22) );
    P( C, D, E, A, B, R(23) );
    P( B, C, D, E, A, R(24) );
    P( A, B, C, D, E, R(25) );
    P( E, A, B, C, D, R(26) );
    P( D, E, A, B, C, R(27) );
    P( C, D, E, A, B, R(28) );
    P( B, C, D, E, A, R(29) );
    P( A, B, C, D, E, R(30) );
    P( E, A, B, C, D, R(31) );
    P( D, E, A, B, C, R(32) );
    P( C, D, E, A, B, R(33) );
    P( B, C, D, E, A, R(34) );
    P( A, B, C, D, E, R(35) );
    P( E, A, B, C, D, R(36) );
    P( D, E, A, B, C, R(37) );
    P( C, D, E, A, B, R(38) );
    P( B, C, D, E, A, R(39) );

#undef K
#undef F

#define F(x,y,z) ((x & y) | (z & (x | y)))
#define K 0x8F1BBCDC

    P( A, B, C, D, E, R(40) );
    P( E, A, B, C, D, R(41) );
    P( D, E, A, B, C, R(42) );
    P( C, D, E, A, B, R(43) );
    P( B, C, D, E, A, R(44) );
    P( A, B, C, D, E, R(45) );
    P( E, A, B, C, D, R(46) );
    P( D, E, A, B, C, R(47) );
    P( C, D, E, A, B, R(48) );
    P( B, C, D, E, A, R(49) );
    P( A, B, C, D, E, R(50) );
    P( E, A, B, C, D, R(51) );
    P( D, E, A, B, C, R(52) );
    P( C, D, E, A, B, R(53) );
    P( B, C, D, E, A, R(54) );
    P( A, B, C, D, E, R(55) );
    P( E, A, B, C, D, R(56) );
    P( D, E, A, B, C, R(57) );
    P( C, D, E, A, B, R(58) );
    P( B, C, D, E, A, R(59) );

#undef K
#undef F

#define F(x,y,z) (x ^ y ^ z)
#define K 0xCA62C1D6

    P( A, B, C, D, E, R(60) );
    P( E, A, B, C, D, R(61) );
    P( D, E, A, B, C, R(62) );
    P( C, D, E, A, B, R(63) );
    P( B, C, D, E, A, R(64) );
    P( A, B, C, D, E, R(65) );
    P( E, A, B, C, D, R(66) );
    P( D, E, A, B, C, R(67) );
    P( C, D, E, A, B, R(68) );
    P( B, C, D, E, A, R(69) );
    P( A, B, C, D, E, R(70) );
    P( E, A, B, C, D, R(71) );
    P( D, E, A, B, C, R(72) );
    P( C, D, E, A, B, R(73) );
    P( B, C, D, E, A, R(74) );
    P( A, B, C, D, E, R(75) );
    P( E, A, B, C, D, R(76) );
    P( D, E, A, B, C, R(77) );
    P( C, D, E, A, B, R(78) );
    P( B, C, D, E, A, R(79) );

#undef K
#undef F

    ctx->state[0] += A;
    ctx->state[1] += B;
    ctx->state[2] += C;
    ctx->state[3] += D;
    ctx->state[4] += E;
}

void sha1_update( sha1_context_t *ctx, const uint8_t *input, uint32_t length )
{
    uint32_t left, fill;

    if( ! length ) return;

    left = ctx->total[0] & 0x3F;
    fill = 64 - left;

    ctx->total[0] += length;
    ctx->total[0] &= 0xFFFFFFFF;

    if( ctx->total[0] < length )
        ctx->total[1]++;

    if( left && length >= fill )
    {
        memcpy( (void *) (ctx->buffer + left),
                (const void *) input, fill );
        sha1_process( ctx, ctx->buffer );
        length -= fill;
        input  += fill;
        left = 0;
    }

    while( length >= 64 )
    {
        sha1_process( ctx, input );
        length -= 64;
        input  += 64;
    }

    if( length )
    {
        memcpy( (void *) (ctx->buffer + left),
                (const void *) input, length );
    }
}

static uint8_t sha1_padding[64] =
{
    0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
};

void sha1_finish( sha1_context_t *ctx, uint8_t digest[SHA1_DIGEST_LEN] )
{
    uint32_t last, padn;
    uint32_t high, low;
    uint8_t msglen[8];

    high = ( ctx->total[0] >> 29 )
         | ( ctx->total[1] <<  3 );
    low  = ( ctx->total[0] <<  3 );

    PUT_UINT32( high, msglen, 0 );
    PUT_UINT32( low,  msglen, 4 );

    last = ctx->total[0] & 0x3F;
    padn = ( last < 56 ) ? ( 56 - last ) : ( 120 - last );

    sha1_update( ctx, sha1_padding, padn );
    sha1_update( ctx, msglen, 8 );

    PUT_UINT32( ctx->state[0], digest,  0 );
    PUT_UINT32( ctx->state[1], digest,  4 );
    PUT_UINT32( ctx->state[2], digest,  8 );
    PUT_UINT32( ctx->state[3], digest, 12 );
    PUT_UINT32( ctx->state[4], digest, 16 );
}

void sha1_hmac_starts( sha1_hmac_context_t *hctx, const uint8_t *key, uint32_t keylen )
{
    uint32_t i;
    uint8_t k_ipad[64];

    memset( k_ipad, 0x36, 64 );
    memset( hctx->k_opad, 0x5C, 64 );

    for( i = 0; i < keylen; i++ )
    {
        if( i >= 64 ) break;

        k_ipad[i] ^= key[i];
        hctx->k_opad[i] ^= key[i];
    }

    sha1_starts( &hctx->ctx );
    sha1_update( &hctx->ctx, k_ipad, 64 );
}

void sha1_hmac_update( sha1_hmac_context_t *hctx, const uint8_t *buf, uint32_t buflen )
{
    sha1_update( &hctx->ctx, buf, buflen );
}

void sha1_hmac_finish( sha1_hmac_context_t *hctx, uint8_t digest[SHA1_DIGEST_LEN] )
{
    uint8_t tmpbuf[SHA1_DIGEST_LEN];

    sha1_finish( &hctx->ctx, tmpbuf );

    sha1_starts( &hctx->ctx );
    sha1_update( &hctx->ctx, hctx->k_opad, 64 );
    sha1_update( &hctx->ctx, tmpbuf, SHA1_DIGEST_LEN );
    sha1_finish( &hctx->ctx, digest );
}

void sha1_hmac( const uint8_t *key, uint32_t keylen, const uint8_t *buf, uint32_t buflen,
                uint8_t digest[SHA1_DIGEST_LEN] )
{
    sha1_hmac_context_t hctx;

    sha1_hmac_starts( &hctx, key, keylen );
    sha1_hmac_update( &hctx, buf, buflen );
    sha1_hmac_finish( &hctx, digest );
}

struct sha1_context_t *sha1_meta()
{
	struct sha1_context_t *c =  (struct sha1_contex_tt *) malloc(sizeof(struct sha1_context_t *));
   	return c;
}

struct sha1_hmac_context_t *sha1_hmac_meta()
{
	struct sha1_hmac_context_t *c =  (struct sha1_hmac_context_t *) malloc(sizeof(struct sha1_hmac_context_t *));
   	return c;
}
